File: python/linear_data_structures/implementation/queue.py
from __future__ import annotations

from collections import deque
from typing import Generic, TypeVar

T = TypeVar("T")


class Queue(Generic[T]):
    def __init__(self) -> None:
        self._items: deque[T] = deque()

    def enqueue(self, item: T) -> None:
        self._items.append(item)

    def dequeue(self) -> T:
        if self.is_empty:
            raise IndexError("dequeue from empty queue")
        return self._items.popleft()

    def peek(self) -> T:
        if self.is_empty:
            raise IndexError("peek from empty queue")
        return self._items[0]

    @property
    def is_empty(self) -> bool:
        return len(self._items) == 0

    @property
    def size(self) -> int:
        return len(self._items)

File: go/linear_data_structures/queue.go
package linear_data_structures

import "container/list"

// Queue is a generic FIFO data structure backed by a linked list.
type Queue[T any] struct {
	items *list.List
}

// NewQueue creates a new empty Queue.
func NewQueue[T any]() *Queue[T] {
	return &Queue[T]{items: list.New()}
}

// Enqueue adds an element to the back of the queue.
func (q *Queue[T]) Enqueue(item T) {
	q.items.PushBack(item)
}

// Dequeue removes and returns the front element. Returns the zero value and
// false if the queue is empty.
func (q *Queue[T]) Dequeue() (T, bool) {
	front := q.items.Front()
	if front == nil {
		var zero T
		return zero, false
	}
	q.items.Remove(front)
	return front.Value.(T), true
}

// Peek returns the front element without removing it. Returns the zero value
// and false if the queue is empty.
func (q *Queue[T]) Peek() (T, bool) {
	front := q.items.Front()
	if front == nil {
		var zero T
		return zero, false
	}
	return front.Value.(T), true
}

// IsEmpty returns true if the queue has no elements.
func (q *Queue[T]) IsEmpty() bool {
	return q.items.Len() == 0
}

// Size returns the number of elements in the queue.
func (q *Queue[T]) Size() int {
	return q.items.Len()
}
